### ISSUE TITLE
HIE unable to find ghc binary for stack project

### ISSUE THREAD

**Issue by wraithm** [haskell/haskell-ide-engine#1504]
I'm on macOS, and I am using haskell-ide-engine with lsp-mode/lsp-haskell in Emacs. I do not have a global ghc binary installed on my machine. I let stack install it locally in the project for each project. HIE is unable to find a ghc binary in the workspace. I had to put the ghc binary into the PATH to get it to work.

**Comment by fendor** [comment]
Thank you for the bug report!
Duplicate of #1495. Fix is in #1496.

**Comment by drewboardman** [comment]
This is also the case for me. Are users expected to have `ghcup` or something installed GHC locally? Is there a way to specify to HIE that I am using `ghc` installed through stack?

**Comment by bubba** [comment]
@drewboardman in your LSP client, invoke hie through `stack exec hie` and add a `hie.yaml` to project to explicitly state that you want to use stack to build your project. There should be a section in the README with an example of the hie.yaml

**Comment by wraithm** [comment]
I tried this, and at least with emacs/lsp-haskell, this did not work. I
also modified the lsp-haskell elisp to be absolutely certain that the right
stack exec command was executed, and I got nothing but errors.

**Comment by fendor** [comment]
@wraithm What errors did you get?
@drewboardman This should work as expected and is a bug that it does not work as expected. As said, this should be fixed in #1496 once it is merged.

**Comment by fendor** [comment]
Fixed by #1496, please try again

**Comment by drewboardman** [comment]
@fendor do we pull and re-install HIE?

Can you give a small explanation of how to check if it is fixed or not? Should I remove my `ghcup` binary of GHC from my path?

**Comment by drewboardman** [comment]
> @drewboardman in your LSP client, invoke hie through `stack exec hie` and add a `hie.yaml` to project to explicitly state that you want to use stack to build your project. There should be a section in the README with an example of the hie.yaml

Like this (using coc.nvim)?

```
  "languageserver": {
    "haskell": {
      "command": "stack exec hie-wrapper",
```

**Comment by fendor** [comment]
@drewboardman Did you add this `hie.yaml` for haskell-ide-engine or for your personal project?
In the personal project, the component is rather `<project>:lib`. You can see the available targets via `stack ide targets`.

And yeah, you should pull from master and pull again. You can prove that it works if you have no ghc on the path, e.g. 
```bash
ghc --version
```
Results in an error, and opening the project works.

**Comment by drewboardman** [comment]
@fendor Yep, pulling and using stack's ghc worked. I do have a question about the `hie.yml` and it's effectiveness. What does putting this in my project do, exactly? It's hard to tell from the readme.

**Comment by fendor** [comment]
Thank you for that feedback, if it is not clear, we must clarify that! 

First, you can select how to build your project, either with stack or cabal. 
Moreover, you can specify some stuff explicitly, e.g. that hie should not try to load certain files because they do not compile or are examples or something.
It also uses another mechanism to compile your project which may be desirable or not.
In general, you specify certain information, like paths to your components, in a way that hie understands them. It is more verbose, but you have more control about what is happening.

**Comment by drewboardman** [comment]
Interesting. So in the example with `haskell-ide-engine:lib` - is there a reason you would place your `hie.yml` in the hie directory, as opposed to your project? Or is that just for the example?

**Comment by fendor** [comment]
I am not sure what you mean.
It is intended that you put your `hie.yaml` (btw, I am pretty sure we are *only* looking for `hie.yaml`) in the root directory of the project, the same directory your `stack.yaml` or `cabal.project` is located.

In `hie.yaml` you specify which build tool to use and which path map to which component, e.g.
```yaml
cradle:
  stack:
    component: "foo:lib"
```
Means that in the current directory, every haskell file is part of the library of the project `foo`. 
You can extend that to multiple components, e.g. tests, benchmarks, executables with:
```yaml
cradle:
  stack:
    - path: src/
      component: "foo:lib"
    - path: test/
      component: "foo:test:unit-tests"
```
This is helpful, because tests and library can have different dependencies.
